File: basic-algorithms/fibonacci/fibonacci.ts
function fibonacci(): () => number {
  let a = 0, b = 1;
  return () => {
    const fib = a;
    [a, b] = [b, a + b];
    return fib;
  }
}

const fib = fibonacci();
console.log(fib()); // 0
console.log(fib()); // 1
console.log(fib()); // 1
console.log(fib()); // 2
console.log(fib()); // 3

File: basic-math/fibonacci/fibonacci.go
package main

import "fmt"

func fibonacci() ()int {
	a,b := 1,0
	return func() int {
		fib := a
		a,b = b,a+b
		return fib
	}
}

func main() {
	f := fibonacci()
	for i := 0; i < 10; i++ {
		fmt.Println(f())
	}
}